src/main/datalog_console/chrome/extension/background/main.cljs

```diff
@@ -0,0 +1,67 @@
+(ns datalog-console.chrome.extension.background.main
+  (:require [goog.object :as gobj]))
+
+
+(defonce tools-conns* (atom {}))
+(defonce remote-conns* (atom {}))
+
+
+(defn handle-devtool-message [devtool-port message _port]
+  (let [tab-id (gobj/get message "tab-id")]
+    (cond
+      (= "init" (gobj/get message "name"))
+      (swap! tools-conns* assoc tab-id devtool-port)
+
+      ;; send message to content-script
+      (gobj/getValueByKeys message "devtool-message")
+      (.postMessage (get @remote-conns* tab-id) message))))
+
+
+(defn set-icon-and-popup [tab-id]
+  (js/chrome.browserAction.setIcon
+   #js {:tabId tab-id
+        :path  #js {"16"  "images/active/icon-16.png"
+                    "32"  "images/active/icon-16.png"
+                    "48"  "images/active/icon-16.png"
+                    "128" "images/active/icon-16.png"}})
+  (js/chrome.browserAction.setPopup
+   #js {:tabId tab-id
+        :popup "popups/enabled.html"}))
+
+
+(defn handle-remote-message [_remote-port message port]
+  (let [tab-id (gobj/getValueByKeys port "sender" "tab" "id")]
+    (cond
+      ; send message to devtool
+      (gobj/getValueByKeys message "datalog-remote-message")
```

Comment:
do you think this should be fully qualified?

Reply:
Tricky to fully qualify this one since it is generated by the user. It is scoped by the port name. We can either:

1. Create a string key on the `.postMessage` inside of `init-datalog-console`
2. Create a namespaced keyword (not to the actual namespace it lives in) but as our reference. However you can't have `#js` keys namespaced. To get around this I introduced a dependency into the extension code for convenience but pushing this onto others doesn't seem the best solution.